scanpipe/pipes/purldb.py

```diff
@@ -320,3 +325,74 @@ def populate_purldb_with_discovered_dependencies(project, logger=logger.info):
         chunk_size=10,
         logger=logger,
     )
+
+
+def send_project_json_to_matchcode(
+    project_json_location, timeout=DEFAULT_TIMEOUT, api_url=PURLDB_API_URL
+):
+    """
+    Given a path to a ScanCode.io Project JSON output, `project_json_location`,
+    send the contents of the JSON file to PurlDB for matching.
+    """
+    project_json_contents = open(project_json_location, "rb")
+    files = {"upload_file": project_json_contents}
+
+    response = request_post(
+        url=f"{api_url}matching/",
+        data={},
+        timeout=timeout,
+        files=files,
+    )
+
+    return response
+
+
+def match_to_purldb(project):
+    """
+    Given a `project` where `scan_output_location` has been set, send the scan
+    of the project to PurlDB for matching. When available, process match results
+    by DiscoveredPackges from the matched package data.
+    """
+    from scanpipe.pipes.d2d import create_package_from_purldb_data
+
+    # send scan to purldb
+    response = send_project_json_to_matchcode(project.scan_output_location)
+    run_url = response["runs"][0]["url"]
+    url = response.get("url")
+    results_url = url + "results/"
+
+    # poll and see if the match run is ready
+    while True:
+        response = request_get(run_url)
+        if response:
+            status = response["status"]
+            if status == "success":
+                break
+        time.sleep(10)
+
+    # get match results
+    match_results = request_get(results_url)
+
+    # map match results
+    matched_packages = match_results.get("packages", [])
+    resource_results = match_results.get("files", [])
+    resource_paths_by_package_uids = defaultdict(list)
+    for matched_package in matched_packages:
+        package_uid = matched_package["package_uid"]
+        for resource in resource_results:
+            if package_uid in resource.get("for_packages", []):
+                resource_paths_by_package_uids[package_uid].append(resource["path"])
+
+    # Map package matches
+    for matched_package in matched_packages:
+        package_uid = matched_package["package_uid"]
+        resource_paths = resource_paths_by_package_uids[package_uid]
+        resources = project.codebaseresources.filter(path__in=resource_paths)
+
+        # Create package matches
+        create_package_from_purldb_data(
+            project,
+            resources=resources,
+            package_data=matched_package,
+            status=flag.MATCHED_TO_PURLDB_PACKAGE,
+        )
```

Comment:
Refactor as functions for easier testing and improved readability.

Reply:
I've ended up splitting `match_to_purldb` into three steps and created tests for each part.